src/java.base/share/classes/java/lang/ClassValue.java

```diff
@@ -369,12 +371,20 @@ private static ClassValueMap getMap(Class<?> type) {
     }
 
     private static final Object CRITICAL_SECTION = new Object();
+    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
     private static ClassValueMap initializeMap(Class<?> type) {
         ClassValueMap map;
         synchronized (CRITICAL_SECTION) {  // private object to avoid deadlocks
             // happens about once per type
-            if ((map = type.classValueMap) == null)
+            if ((map = type.classValueMap) == null) {
                 type.classValueMap = map = new ClassValueMap();
+                // Place a Store fence as the last operation to emulate
+                // ClassValueMap containing final fields. This ensures it can be
+                // published safely in the non-volatile field Class.classValueMap,
+                // since stores to the fields of ClassValueMap will not be reordered
+                // to occur after the store to the field type.classValueMap
+                UNSAFE.storeFence();
```

Comment:
Wait a second, how's that supposed to work? The `storeFence` should be between the end of `new ClassValueMap()` and the store to `classValueMap`.

Reply:
Ah yes, my bad 🤦‍♂️

Reply:
Updated PR.

Reply:
Now I wonder (after reading the discussion), why not `UNSAFE.storeStoreFence()`? I am basically okay with either, given they currently map to the same thing. But it seems that semantically only `storeStoreFence` is really needed here.

Reply:
Scratch that. Hotspot does `MemBar_Release` for final stores, low-level Java code is better do the same:
 https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/parse1.cpp#L1005

It deserves a separate follow-up to homogenize the uses of `Unsafe.storeStoreFence` and `Unsafe.storeFence` in other places in JDK.

Reply:
@galderz did you really update the PR? I still see the _storeFence_ before the write.

Reply:
ah, sorry I got it now. Please ignore my previous comment.